### PR TITLE
Upgradeable HATVaults

### PR DESCRIPTION
**Pull request by ben-kaufman**
Prepare hat vaults for upgradeability

### CODE REVIEW

File: .openzeppelin/unknown-31337.json [hunk 1]
```diff
@@ -0,0 +1,2222 @@
+{
```
Comment (jellegerbrandy) [inline .openzeppelin/unknown-31337.json:1]: committed by mistake?

File: contracts/vaults/Claim.sol [hunk 2]
```diff
@@ -4,7 +4,7 @@ pragma solidity 0.8.6;
 import "./Base.sol";
 
 contract Claim is Base {
-    using SafeERC20 for IERC20;
+    using SafeERC20Upgradeable for IERC20Upgradeable;
```
Comment (jellegerbrandy) [inline contracts/vaults/Claim.sol:7]: why? we are calling that token, we do not care if it upgradeable or not, or do we?

File: contracts/vaults/Pool.sol [hunk 3]
```diff
@@ -65,7 +65,7 @@ contract Pool is Base {
             }));
         }
         poolInfos.push(PoolInfo({
-            lpToken: IERC20(_lpToken),
+            lpToken: IERC20Upgradeable(_lpToken),
```
Comment (jellegerbrandy) [inline contracts/vaults/Pool.sol:68]: cam just be IERC20

File: scripts/hatvaultsdeploy.js [hunk 4]
```diff
@@ -1,77 +1,84 @@
...
-  console.log("Deploying the contracts with the account:", deployerAddress);
-  console.log("Account balance:", (await deployer.getBalance()).toString());
+  if (!silent) {
+    console.log("Deploying the contracts with the account:", deployerAddress);
+    console.log("Account balance:", (await deployer.getBalance()).toString());
+  }
 
   const HATVaults = await ethers.getContractFactory("HATVaults");
-  const hatVaults = await HATVaults.deploy(
+  const hatVaults = await upgrades.deployProxy(HATVaults, [
```
Comment (jellegerbrandy) [inline scripts/hatvaultsdeploy.js:39]: we should also decide on which proxy pattern to use: transparent or uups

It's a setting: https://docs.openzeppelin.com/upgrades-plugins/1.x/api-hardhat-upgrades

**Review by jellegerbrandy — CHANGES_REQUESTED** [review]
some small changes.

@ben-kaufman do you have an opinion on transparentproxy vs the uups one?

**Comment by ben-kaufman** [comment]
> @ben-kaufman do you have an opinion on transparentproxy vs the uups one?
UUPS is nice but it will make our HATVaults even heavier. Right now it's already about 1MB too big to deploy. So unless we can significantly reduce contract size we don't really have the option of UUPS.

File: contracts/vaults/Pool.sol (hunk 3 above)
Comment (ben-kaufman) [reply contracts/vaults/Pool.sol:68]: It gives an error deploying:
Error: Contract `HATVaults` is not upgrade safe

@openzeppelin/contracts/utils/Address.sol:191: Use of delegatecall is not allowed

Also even if can ignore the error I think it's best practice to use only the upgradeable repo with upgradeable contracts

Comment (jellegerbrandy) [reply contracts/vaults/Pool.sol:68]: The error must be a bug, right?  In any case, I appreciete your point about using just a single dependency

**Pull request merged**